### ISSUE TITLE
Add context menu items to facilitate editing

### ISSUE THREAD

**Issue by cpcallen** [rachel-fenichel/blockly-keyboard-experimentation#132]
Add the following items to the context menus for blocks, with the the given keyboard shortcuts:

* [x] **Edit contents** (`→`): Moves the cursor (focus) to the first focusable field or input.
  * This option exists mainly to facilitate discoverability of the right arrow as a means to navigate to the contents of a block.  It should probably be the first item in the context menu.
* [ ] **Move block** (`m`): "Picks up" the block to move it using the dragging workflow (see #106).
* [x] #173
  * See #102 for details of how this should work.
* [x] **Delete block** (`del`+`backspace`): delete the block, as current delete shortcut does.
* [x] [**Cut** (`⌘X` etc)](https://github.com/google/blockly-keyboard-experimentation/issues/217)
* [x] [**Copy** (`⌘X` etc)](https://github.com/google/blockly-keyboard-experimentation/issues/218)
* [x] [**Paste** (`⌘X` etc): ](https://github.com/google/blockly-keyboard-experimentation/issues/219)Paste (see #128).

Add the following item to the context menu for unconnected next and input connections, as well as when the cursor is visiting the workspace:

* [ ] **Place block** (same as replace block, above): place a new block at this location using the marker workflow.

The names given above are not necessarily intended to be the final English names for the given items, merely descriptive.

**Comment by cpcallen** [comment]
@rachel-fenichel: as I [noted in the discussion with Kirsty](https://github.com/google/blockly-keyboard-experimentation/issues/130#issuecomment-2605515323), I think "Insert" should be the primary shortcut for adding blocks, and I'm not sure that if we have insert we need replace.

**Comment by kmcnaught** [comment]
Is this a question about naming, shortcut key choice, or functionality? Or all three?

If you are at a prev/next connection then "insert" would add a new block in the current location. 

If you were at an in/out connection (_not_ a block) then an action called "insert" might replace the currently-connected block(s) with the newly selected one. Are there any geras-style platforms where it might be appropriate/meaningful to insert a block between in/out as an extra block, without replacing anything? I don't think this would ever be the case in MakeCode.

If you were at a [single cursor location that collapses a block and its out connection] then I was imagining an action called "replace" would replace the currently-connected block(s) with the newly selected one. 

What we can't rely on is using "delete" followed by "insert" to work for reporter blocks as an alternative for "replace", since deleting might (nearly always in MakeCode) leave you with a shadow block. So for blocks connected at an in/out connection, there needs to be an action that is semantically equivalent to "replace" even if you decide it's unambiguous to call it "insert" (I think it's much easier to argue this at an in/out connection than at a "collapsed single location".)

For instance, if we collapse to a single cursor location (as per Matt's demo) then in the below situation, I would expect a context menu to give me a single action for "insert/replace" that replaces the block. 
![screenshot of pause blocks, one with shadow block and one with variable](https://github.com/user-attachments/assets/9e6ca683-62b7-40c6-9c16-28f531e1d426)

I can see an argument for keeping the naming/shortcut keys equivalent for these actions, and I can't see a situation where "Insert" and "Replace" would be 2 different valid actions, _unless_ there's some geras-style applications where inserting in the middle of a horizontal stack of blocks would be desirable

**Comment by rachel-fenichel** [comment]
We discussed this live today with @cpcallen, @microbit-matt-hillsdon, Lucy, Emma, and Robert. I _think_ our conclusions were:
- All of the insertion/replacement flows you described need to be possible (so we seem to agree about required functionality)
- The name of the action, as shown to the user, may be different in different places (e.g. insert vs replace)
- Replacement is more likely in horizontal cases, while insertion is more likely in vertical cases
- Christopher objects to having both "Insert (above)" and "Replace" on a stack block, finding them redundant
- Many of these cases are handled in core Blockly in the dragging logic, and Christopher and Rachel think we should match that logic where possible
  - Rachel also wants this because it means free functionality by just calling `connect`

In https://github.com/google/blockly-keyboard-experimentation/pull/196 I implemented the necessary logic for replacing a reporter block (block with an output) with an appropriate block from the flyout/toolbox.

I agree that in your case above, the context menu has an insert/replace action.

**Comment by rachel-fenichel** [comment]
> Are there any geras-style platforms where it might be appropriate/meaningful to insert a block between in/out as an extra block, without replacing anything? I don't think this would ever be the case in MakeCode.

Yes! And I can make it happen in MakeCode too. This feature is called "healing a stack" or "splicing". It's implemented in [`unplugFromRow`](https://github.com/google/blockly/blob/develop/core/block.ts#L420) and [`unplugFromStack`](https://github.com/google/blockly/blob/develop/core/block.ts#L495) in core Blockly.

### Examples
In the [advanced Blockly playground](https://blockly-demo.appspot.com/static/tests/playgrounds/advanced_playground.html#renderer=geras&rtl=false&debugEnabled=false), with our basic blocks:
![Image](https://github.com/user-attachments/assets/fdd409e4-bc86-4478-861b-21bd37ec0e3d)

And the same thing in MakeCode:

![Image](https://github.com/user-attachments/assets/4710a943-34fa-4c6b-8f99-b31aabee37da)

### Explanation
Situations that trigger this:
 - You have a parent **block A** with a single value input (print block) 
![Image](https://github.com/user-attachments/assets/d949a974-f7d4-4ea5-b6b1-bbfea6d20e7b)

 - There is a connected **block B** with a single value output (number block: 3.14)
![Image](https://github.com/user-attachments/assets/a1a6d392-adb0-4cc8-80e9-f53eaffad131)

 - There is a **block C** that has a single value input and a single value output, both of the appropriate type (square root block)
![Image](https://github.com/user-attachments/assets/5554fc1b-e89b-4f5a-a2eb-af955c5a9aa2)

 - Block **C** is dragged between blocks **A** and **B**
 - When the drag ends, the blocks will be connected horizontally in order (from left to right) **A**, **C**, **B**
![Image](https://github.com/user-attachments/assets/433602cc-5539-4b38-a479-9a5be3cd9f84)

 
Note that in both cases there is a shadow **block D** that is initially attached to **block C**'s input. Block **D** disappears, because block **B** covers it up. 
![Image](https://github.com/user-attachments/assets/07802a15-a49d-43fd-b4a5-59ccd879d94e)

### Other info

You cannot trigger a splice if block **C** has two value inputs, because the correct operation is ambiguous:

Geras:
![Image](https://github.com/user-attachments/assets/77a08948-36c9-4988-abb8-f3c71d5f7c86)

Zelos:

![Image](https://github.com/user-attachments/assets/c112b76f-c1dd-42ba-bf72-b7d0d97784bc)

You **can** trigger a splice if block **B** has two value inputs, because the correct operation is unambiguous:

Geras:

![Image](https://github.com/user-attachments/assets/f495a3c8-6cb1-43e6-92a4-470d883824d2)

Zelos:
![Image](https://github.com/user-attachments/assets/4ead559c-c8ad-4667-8437-6a015f31bf8d)

**Comment by kmcnaught** [comment]
Ooh. That is what I had pictured in geras, but a bit surprising in MakeCode. Thanks for the examples! 

My initial position is that we would expect the same behaviour from a keyboard-triggered "insert/replace" operation as you would get from a mouse drag. Now I understand more I can imagine that sometimes it's going to be a bit awkward if the splicing isn't what was intended. I don't think it changes my views on the core insert/replace flow, just a slightly surprising artefact of existing behaviour. 

Do you have any intention to give more control over this, or does it change your views at all?

**Comment by rachel-fenichel** [comment]
> I don't think it changes my views on the core insert/replace flow, just a slightly surprising artefact of existing behaviour.

Same here, especially because it lets us rely on existing (and well-tested) code in core, rather than throwing together a new layer of complex checks.

Please give insertion/replacement for inputs and outputs a try using the T key after #196 and let me know if it feels reasonable.

That change does not address the issues of when insert/replace should be in the context menu or what the text of the insert/replace option should be.

**Comment by kmcnaught** [comment]
I've just tried it. The general feel of inserting with 't' is very reasonable and aligned with what I was hoping for. 

There's a slight weirdness that you end up with the cursor at the connection (which is usually skipped) after inserting. I don't know if this will lead to any bugs in the context menu etc, or if it's just a cosmetic bug. 

I can imagine that once copy/paste is working, this will all feel quite powerful. It would be easy to e.g. remove the `sin` from the outside of a number block by copying/cutting the child, deleting the parent, and pasting the child in again. 

I still don't quite understand the rules around splicing. There's a [30 second video here](https://www.loom.com/share/ad28d2470fc84f2da5abf4233aa3adc8?sid=c8013499-aee5-4305-9b15-0864c321594e) showing 2 situations where the splicing behaviour is different but I expected the same. 

(to clarify - I don't think we should _change_ splicing behaviour for keyboard nav, unless there are bug fixes that apply everywhere - I'm just getting my head around the logic)

**Comment by rachel-fenichel** [comment]
Yes, that behaviour is consistent with the other behaviour but is also confusing in the zelos case. If we change it for keyboard nav we may also want to change it for mouse. (In which case would file that as a separate bug to figure out the exact conditions for a different behaviour.)

Explanation:

Here's a gif of the same flow with two changes: I turned on the geras renderer and set the theme to the standard Blockly theme, which renders shadow blocks as lighter versions of real blocks, to make it easier to see the differences.

![Image](https://github.com/user-attachments/assets/2f764b79-4fc0-46d6-bb6a-1e22e75627da)


The behaviour is confusing because it relies on the distinction between shadow blocks and real blocks, which is minimal in zelos. 

The `x + y` block is a real block, with two shadow children (`1` and `4`).

The `sin` block is a real block with one shadow child (`45`).

The `repeat X times` block is a real block with one shadow child (`10`).
 
Changing the value in the number block to `123` does not convert it to a real (non-shadow) block. There is a [plugin](url) that does convert shadow blocks to real blocks on edit.

When you insert the `sin` block at the output of the `x + y` block, it splices because the real `x + y` block is higher priority than the shadow number block (`45`) that was already attached to the `sin` block.

When you insert the `sin` block at the output of the `123` block it replaces because the shadow `123` is lower priority than the shadow `45` that was already attached to the `sin` block.

At the end of the gif I navigate to each `sin` block and delete it. Notice that the `123` comes back for the lower `repeat`: when a shadow block is obscured, its value is stashed and reapplied if the obscuring block is deleted. That's also why the `10` block reappears as a `10`.

**Comment by kmcnaught** [comment]
Thanks for the explanation and demo, that makes a lot of sense. I can't help but favour the plugin's behaviour of turning a shadow block into a real block on edit - since that shows intent from the user to make something concrete. Are there situations or applications in which this wouldn't make sense, or previous discussions I can refer back to learn about the competing needs?

There are downstream consequences of this behaviour on copy and paste, which don't make up with my expectations. I would have assumed you'd be able to copy and paste the contents of an oval, like the text in a "write with shadow" block

![Screenshot. Write with shadow (item variable). Write with shadow "how are you". The block containing "how are you" is selected. ](https://github.com/user-attachments/assets/9a374b70-2e66-4354-b0e1-c55b140a9289)

I would naively argue that these blocks should be real, and draggable, and copyable. If they are not, then we definitely need some user feedback when they try to copy a block, because currently copy will silently fail and you'll end up surprised by pasting the last thing in the copy buffer.

**Comment by rachel-fenichel** [comment]
The main reason to have shadow blocks is that they are undeletable, so you can never end up with an empty input.

Converting it to a real block on change means that the edited block would get bumped out, instead of obscured, by dragging on a new block.

The issue with not being able to copy them is due to the precondition checks for copying and should be resolvable, so I've added an issue for that.

**Comment by rachel-fenichel** [comment]
Remaining work not required for testing, so this is now in the "simple editing complete" milestone.

**Comment by kmcnaught** [comment]
I have re-read this thread after getting a bit more firsthand experience of the splicing behaviour, and I'm losing confidence in the current experience of where blocks are pasted, when you're on a statement block. I discussed a specific situation in #302.

**Comment by maribethb** [comment]
Closing since most is done and we have tracking for the smaller issues now.
